### ISSUE TITLE
gemma3-4b 不能正常转换/运行

### ISSUE THREAD

**Issue by kuku20009** [airockchip/rknn-llm#240]
新版本1.2.0转换gemma3-4b会出现：
ERROR: Catch exception when loading model: 'Gemma3Config' object has no attribute 'vocab_size'

用unsloth的版(https://huggingface.co/unsloth/gemma-3-4b-it)本替换可以正常转换但是运行异常：

![Image](https://github.com/user-attachments/assets/a5a7852d-993f-4ac3-acd4-9a0b0aa6adb7)

不管加不加提示词模板：
<start_of_turn>user\n
<end_of_turn>\n<start_of_turn>model\n

输出都是下面的：

![Image](https://github.com/user-attachments/assets/e7b03ca0-ad94-43f3-98f5-c41fdfec9648)

**Comment by c0zaut** [comment]
I need to update my version of the Gradio app, which just uses token input and is a lot easier to deal with. Unfortunately, they force text input in multimodal by requiring a `<image>` placeholder tag.

In the meantime, try changing this section: https://github.com/airockchip/rknn-llm/blob/cb5b341364311065fd19eddd631a79a9f0c5afe1/examples/rkllm_server_demo/rkllm_server/gradio_server.py#L225

To this. 
```python
        system_prompt = ""
        prompt_prefix = "<bos><start_of_turn>user\nYou are a helpful assistant.\n\n"
        prompt_postfix = "<end_of_turn>\n<start_of_turn>model"
        self.set_chat_template(self.handle, ctypes.c_char_p(system_prompt.encode('utf-8')), ctypes.c_char_p(prompt_prefix.encode('utf-8')), ctypes.c_char_p(prompt_postfix.encode('utf-8')))
```

And then play around with the temperature, top_p, repeat_penalty, etc settings here: https://github.com/airockchip/rknn-llm/blob/cb5b341364311065fd19eddd631a79a9f0c5afe1/examples/rkllm_server_demo/rkllm_server/gradio_server.py#L191

Due to how the models are quantized, I generally avoid top_k that isn't 1; that being said, it's good to have an "anchor" parameter and tweak individual settings one-by-one to get the desired effect.

**Comment by c0zaut** [comment]
![Image](https://github.com/user-attachments/assets/cdbf1daa-4659-43ff-a7e6-ebac746852dc)

**Comment by kuku20009** [comment]
@c0zaut please help me

1、first try your code to convert rkllm:
modelpath = '/data/qwen/gemma-3-4b-it'

from transformers import Gemma3Processor, Gemma3ForConditionalGeneration
model = Gemma3ForConditionalGeneration.from_pretrained(modelpath, device_map='cpu').eval()
processor = Gemma3Processor.from_pretrained(modelpath, use_fast=True)
model.language_model.save_pretrained('llm')
processor.tokenizer.save_pretrained('llm')
del model
del processor

from rkllm.api import RKLLM
gemma3 = RKLLM()
gemma3.load_huggingface('llm')

gemma3.build(do_quantization=True, optimization_level=0, quantized_dtype='w8a8', quantized_algorithm='normal', target_platform='rk3588', num_npu_core=3, extra_qparams=None, dataset=None, hybrid_rate=0, max_context=4096 * 4)
gemma3.export_rkllm('llm/gemma3text.rkllm')

2、then mod gradio_server.py
system_prompt = ""
prompt_prefix = "<bos><start_of_turn>user\nYou are a helpful assistant.\n\n"
prompt_postfix = "<end_of_turn>\n<start_of_turn>model"
self.set_chat_template(self.handle, ctypes.c_char_p(system_prompt.encode('utf-8')), ctypes.c_char_p(prompt_prefix.encode('utf-8')), ctypes.c_char_p(prompt_postfix.encode('utf-8')))

mod top_k=40 tok_p=0.95

3、still get wrong response
![Image](https://github.com/user-attachments/assets/0c08958f-c0b1-4063-9a38-fbf8ad4d6cfc)

[gradio_server.py.txt](https://github.com/user-attachments/files/19774538/gradio_server.py.txt)

**Comment by c0zaut** [comment]
Have you tried past the first turn? For some reason, the first response is usually a throw-away, but even with chat history disabled, the following responses are correct. It also works with history enabled. Using the original Gemma3 model from Google, and default chat template, I get the following results. This is on a version that I aligned with multimodal data (still need to add image input for testing, though):

![Image](https://github.com/user-attachments/assets/a7a287fd-6e78-4228-9047-497c580b3d3f)

![Image](https://github.com/user-attachments/assets/e9681a8c-e140-436a-976a-cac67a027e75)

![Image](https://github.com/user-attachments/assets/2feda661-3a7d-4278-a446-d1679a17e658)

So there's something weird in the initial "warm up" stage that's going on. 

@waydong   - any custom config suggestions for the conversion?

**Comment by c0zaut** [comment]
I think there's some funky going on with its internal cache. This is with history disabled. Going to try a groupwise quant and see if that improves accuracy, at the cost of reduced speed. Context is 16K, but none of the parameters really seem to make a difference.

![Image](https://github.com/user-attachments/assets/b79ceb25-96c2-4430-be4c-cb53691516a5)

![Image](https://github.com/user-attachments/assets/d28e6ce3-02f7-4bb9-ae4d-5a27379afa94)

![Image](https://github.com/user-attachments/assets/9138f550-73f0-460e-84b8-e480128bd445)

![Image](https://github.com/user-attachments/assets/42361798-b2eb-49cc-9083-18e8108e808c)

![Image](https://github.com/user-attachments/assets/2483e934-4594-4a52-9290-c6b3c92aef79)

**Comment by waydong** [comment]
Have you tried using the original Hugging Face model for pure text-based Q&A to see if it works properly?

**Comment by c0zaut** [comment]
I'm using the official instruct 4B version from Google HF and exporting the text-based model. 

I just tried it with full g128, default chat template, and that worked, but it's much slower:

![Image](https://github.com/user-attachments/assets/79bf168c-6c11-4498-a533-280c17e8be81)

![Image](https://github.com/user-attachments/assets/8483e621-0cc0-4ae4-96f5-eddffe578b68)

Conversion script:

```python
from transformers import Gemma3Processor, Gemma3ForConditionalGeneration
import safetensors
import torch

# Gated repo - requires HF_TOKEN environment variable to access
modelpath = 'google/gemma-3-4b-it'

model = Gemma3ForConditionalGeneration.from_pretrained(modelpath, device_map='auto', torch_dtype=torch.bfloat16).eval()
processor = Gemma3Processor.from_pretrained(modelpath, use_fast=True)

model.language_model.save_pretrained('llm')
processor.save_pretrained('llm')

modelpath = 'llm'
savepath = 'llm/gemma-3-4b-it-g128.rkllm'

llm = RKLLM()

ret = llm.load_huggingface(model=modelpath, device='cuda')
if ret != 0:
    print('Load model failed!')
    exit(ret)


ret = llm.build(
        do_quantization=True, 
        optimization_level=0, 
        quantized_dtype='w8a8_g128', 
        max_context=4096 * 4,
        quantized_algorithm='normal', 
        target_platform='rk3588', 
        num_npu_core=3, 
        extra_qparams=None, 
        dataset=None
        )
if ret != 0:
    print('Build model failed!')
    exit(ret)


ret = llm.export_rkllm(savepath)
if ret != 0:
    print('Export model failed!')
    exit(ret)
```

**Comment by c0zaut** [comment]
I also have a method for exporting the vision head as a single model, as well, but performance is not quite where I want it to be since the attention is single threaded, so it takes about 40s to run (30s of it is exSDPA.)

**Comment by c0zaut** [comment]
Guess I will need to play around with alignment data and the chat template, as well, for w8a8, since that is much more sensitive. For multimodal input, would it be possible to use tokens and embeds, as well as the current implementation? I have found I generally get better results when I just token inputs, although I have not tested that with the latest version.

**Comment by c0zaut** [comment]
Update - hybrid ratio of 0.25 with w8a8 quantization seems to be the best way to go, at least for solving this initial response issue without too much of a speed hit. I'm going to try a modified chat template to see if that improves things with vanilla w8a8, since optimization with 30 examples seems to have absolutely no effect. Hybrid quant takes much longer for conversion, though. Without optimization, it's about an hour and a half using an i7 CPU and 64GB of DDR4 RAM. With an optimized dataset of 30, there's another half hour added, for a total of two hours on pretty decent hardware. 

![Image](https://github.com/user-attachments/assets/ffd79774-c80d-47ec-b9b7-7f755a07aeb8)

![Image](https://github.com/user-attachments/assets/9ab6e644-e7fc-4708-b584-e7529696eaca)

```python
=========init....===========
I rkllm: rkllm-runtime version: 1.2.0, rknpu driver version: 0.9.8, platform: RK3588
I rkllm: loading rkllm model from gemma-3-4b-it-0.25.rkllm
I rkllm: rkllm-toolkit version: 1.2.0, max_context_limit: 16384, npu_core_num: 3, target_platform: RK3588
I rkllm: Enabled cpus: [4, 5, 6, 7]
I rkllm: Enabled cpus num: 4
I rkllm: system_prompt: 
I rkllm: prompt_prefix: <start_of_turn>user\n
I rkllm: prompt_postfix: <end_of_turn>\n<start_of_turn>model\n
RKLLM Model has been initialized successfully！
==============================
```

Output isn't that great, but it's at least mostly coherent.

**Comment by c0zaut** [comment]
@kuku20009 

Be warned - conversion takes a VERY long time due to hybrid quantization (w8a8 75%, w8a8_g512 25%). It takes about an hour and a half without optimization on an i7.

```python
from transformers import Gemma3Processor, Gemma3ForConditionalGeneration
import safetensors
import torch

# Gated repo - requires HF_TOKEN environment variable to access
modelpath = 'google/gemma-3-4b-it'

model = Gemma3ForConditionalGeneration.from_pretrained(modelpath, device_map='cpu', torch_dtype=torch.bfloat16).eval()
processor = Gemma3Processor.from_pretrained(modelpath, use_fast=True)

model.language_model.save_pretrained('llm')
processor.save_pretrained('llm')

del model
model = None
del processor
processor = None

modelpath = 'llm'
savepath = 'llm/gemma-3-4b-it-g128.rkllm'

llm = RKLLM()

ret = llm.load_huggingface(model=modelpath, device='cpu')
if ret != 0:
    print('Load model failed!')
    exit(ret)


ret = llm.build(
        do_quantization=True, 
        optimization_level=0, 
        quantized_dtype='w8a8', 
        # hybrid ratio of 25% gives a good balance
        hybrid_rate=0.25,
        max_context=4096 * 4,
        quantized_algorithm='normal', 
        target_platform='rk3588', 
        num_npu_core=3, 
        extra_qparams=None, 
        dataset=None
        )
if ret != 0:
    print('Build model failed!')
    exit(ret)


ret = llm.export_rkllm(savepath)
if ret != 0:
    print('Export model failed!')
    exit(ret)
```

For inference, I used these settings, testing translating Chinese to English. It works very well with the cache enabled. 
```python
        rkllm_param.max_context_len = 4096 * 4
        rkllm_param.max_new_tokens = 4096 * 2
        rkllm_param.skip_special_token = True
        rkllm_param.n_keep = -1
        rkllm_param.top_k = 64
        rkllm_param.top_p = 0.95
        rkllm_param.temperature = 0.7
        rkllm_param.repeat_penalty = 1.0
        rkllm_param.frequency_penalty = 1.0
        rkllm_param.presence_penalty = 0.0

        rkllm_param.mirostat = 0
        rkllm_param.mirostat_tau = 5.0
        rkllm_param.mirostat_eta = 0.1

        rkllm_param.is_async = True
```

Chat template:

```python
        system_prompt = ""
        prompt_prefix = "<start_of_turn>user\nYou are a helpful assistant.\n"
        prompt_postfix = "<end_of_turn>\n<start_of_turn>model\n"
        self.set_chat_template(
            self.handle, 
            ctypes.c_char_p(system_prompt.encode('utf-8')), 
            ctypes.c_char_p(prompt_prefix.encode('utf-8')), 
            ctypes.c_char_p(prompt_postfix.encode('utf-8')))
```

To clear the cache, I added the following code to the bottom of the RKLLM class's init function

```python
        self.rkllm_release_prompt_cache = rkllm_lib.rkllm_release_prompt_cache
        self.rkllm_release_prompt_cache.argtypes = [RKLLM_Handle_t]
        self.rkllm_release_prompt_cache.restype = ctypes.c_int

        self.rkllm_clear_kv_cache = rkllm_lib.rkllm_clear_kv_cache
        self.rkllm_clear_kv_cache.argtypes = [RKLLM_Handle_t, ctypes.c_int]
        self.rkllm_clear_kv_cache.restype = ctypes.c_int
```

Added this function to the class:

```python
    def clear_kv_cache(self, keep_sys_prompt=0):
        ret = self.rkllm_clear_kv_cache(self.handle, ctypes.c_int(keep_sys_prompt))
        if ret != 0:
            print(f"Clear KV cache failed! {ret}")
        if self.prompt_cache_path != None:
            ret = self.rkllm_release_prompt_cache(self.handle)
            if ret != 0:
                print(f"Release prompt cache failed! {ret}")
        return
```

And then updated the clear button's click action accordingly:

```python
clear.click(lambda: None, inputs=None, outputs=rkllmServer, queue=False).then(rkllm_model.clear_kv_cache, None, rkllmServer)
```

**Comment by kuku20009** [comment]
@c0zaut thank you very much !
